File: app/src/main/java/com/billpayments/features/saved_billers/data/SavedBillersStore.kt
package com.billpayments.features.saved_billers.data

import com.billpayments.core.event.UiEvent
import com.billpayments.core.event.UiEventBus
import com.billpayments.features.auth.data.AuthStore
import com.billpayments.features.auth.data.userIdOrNull
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

@Singleton
class SavedBillersStore @Inject constructor(
    authStore: AuthStore,
    private val repository: SavedBillersRepository,
    private val events: UiEventBus,
    private val scope: CoroutineScope,
) {
    private var epoch = 0
    private var userId: String? = null
    private val _state = MutableStateFlow(initialState(authStore))
    val state: StateFlow<SavedBillers> = _state.asStateFlow()

    init {
        scope.launch {
            authStore.state.map { it.userIdOrNull }.distinctUntilChanged().collect { id ->
                epoch++
                userId = id
                _state.value = if (id == null) SavedBillers.EMPTY else repository.restore(id)
            }
        }
    }

    private fun initialState(authStore: AuthStore): SavedBillers {
        val id = authStore.state.value.userIdOrNull ?: return SavedBillers.EMPTY
        userId = id
        return repository.restore(id)
    }

    fun save(saved: SavedBiller) {
        if (_state.value.contains(saved.billerId, saved.account)) return
        commit(_state.value.adding(saved))
    }

    fun remove(billerId: String, account: String) {
        commit(_state.value.removing(billerId, account))
    }

    private fun commit(next: SavedBillers) {
        val id = userId ?: return
        val previous = _state.value
        val startedEpoch = epoch
        _state.value = next
        scope.launch {
            try {
                repository.persist(id, next)
            } catch (_: Exception) {
                if (startedEpoch != epoch) return@launch
                _state.value = previous
                events.emit(UiEvent.StorageFailed)
            }
        }
    }
}

File: BillPayments/Features/SavedBillers/Data/SavedBillersStore.swift
import Observation

@Observable
final class SavedBillersStore {
    private(set) var state: SavedBillers
    @ObservationIgnored private(set) var pendingWork: Task<Void, Never>?
    @ObservationIgnored private(set) var observationWork: Task<Void, Never>?

    @ObservationIgnored private let repository: SavedBillersRepository
    @ObservationIgnored private let events: UiEventBus
    @ObservationIgnored private var epoch = 0
    @ObservationIgnored private(set) var userId: String?

    init(
        authStore: AuthStore,
        repository: SavedBillersRepository,
        events: UiEventBus
    ) {
        self.repository = repository
        self.events = events
        if let id = authStore.state.userIdOrNull {
            self.userId = id
            self.state = repository.restore(userId: id)
        } else {
            self.state = .empty
        }
        observationWork = Task { [weak self, authStore] in
            for await userId in Observations({ authStore.state.userIdOrNull }) {
                guard let self else { return }
                self.handleAuthChange(userId)
            }
        }
    }

    private func handleAuthChange(_ id: String?) {
        guard id != userId else { return }
        epoch += 1
        userId = id
        state = id.map { repository.restore(userId: $0) } ?? .empty
    }

    func save(_ saved: SavedBiller) {
        if state.contains(billerId: saved.billerId, account: saved.account) { return }
        commit(state.adding(saved))
    }

    func remove(billerId: String, account: String) {
        commit(state.removing(billerId: billerId, account: account))
    }

    private func commit(_ next: SavedBillers) {
        guard let id = userId else { return }
        let previous = state
        let startedEpoch = epoch
        state = next
        pendingWork = Task {
            do {
                try await repository.persist(userId: id, value: next)
            } catch {
                if startedEpoch != epoch { return }
                state = previous
                events.emit(.storageFailed)
            }
        }
    }
}
